services/transientsource/transientsource.go

```diff
@@ -0,0 +1,140 @@
+package transientsource
+
+import (
+	"context"
+	"encoding/json"
+	"sync"
+
+	backendconfig "github.com/rudderlabs/rudder-server/config/backend-config"
+	"github.com/rudderlabs/rudder-server/jobsdb"
+	"github.com/rudderlabs/rudder-server/utils/pubsub"
+	"github.com/tidwall/gjson"
+)
+
+// Service provides services related to transient source ids
+type Service interface {
```

Comment:
Another software design dilemma for you :sweat_smile: 

Again, your code does what is supposed to do but I'd like to point out that declaring interfaces named after "things" instead of "actions" is kind of a bad practice.

That's why we usually see interfaces in the Go standard library end with "er" (e.g. `Server`, `Reader`, `Writer`, `Closer`, etcetera) and not `Service`.

The main reason is that if you name interfaces after things they'll notoriously get big (or at least bigger than their "action" couterpart and less understable) and give birth to fat objects with multiple responsibilities (see SOLID principles).

In fact we see packages using this `Service` that don't even need the whole set of methods (i.e. the `batchrouter` which is just using `Apply(sourceID)`) so now we're under the impression that the `batchrouter` needs the whole service and we don't even know what a service is/does unless we look at the whole thing.

FYI, interfaces can also be embedded like `Reader` + `Closer` = [ReadCloser](https://pkg.go.dev/io#ReadCloser), so if some object is in need of more "doers" in a single abstraction it can still be done like in the standard library.

Again, not a blocking issue but please do keep in mind that these kind of interfaces are usually a code smell.

Reply:
Yes, I have read [the bible](https://go.dev/doc/effective_go) and I know how I should name single-method interfaces, I also know how we should combine and name 2 single method interfaces in one. But should 2 be the limit for the number of an interface's methods? And should everything be named using -er? No, I don't believe that naming an interface after a [thing](https://github.com/prometheus/client_golang/blob/main/prometheus/histogram.go#L46) is a bad practice.

Why is it better to have a single struct implementing 5 single-method interfaces, instead of a struct implementing a single 5-method interface (can use embedding there as well) when these methods are related to each other?

Is [this design](https://github.com/rudderlabs/rudder-server/blob/69857216e8b736d2924a761a8eef578edce8d40e/app/cluster/dynamic.go#L42) better? In this scenario, we are declaring the `lifecycle` interface (a.k.a. `StartStopper`) as we only need to use 2 methods inside `Dynamic`. But in the `Dynamic` struct, not all `lifecycle` fields are made equal. Therefore, we are forced to use the field's name as a means to indicate the actual type of the field that we expect, e.g. in `GatewayDB`  we need a `jobsdb.HandleT`, in `Processor` we expect  `processor.HandleT` etc.
With the above design I would argue that by using the small `lifecycle` interface for everything we essentially give away many of the benefits provided by strong typing.